// @oasislabs/client/dist/lib/index.js
import { keccak256 } from 'js-sha3';
import Gateway from '@oasislabs/gateway';
import { cbor, bytes } from '@oasislabs/common';
import { OasisCoder, Service, Address, Balance, deploy, DeployHeader, setGateway, defaultOasisGateway, fromWasm, fromWasmSync, } from '@oasislabs/service';
import { Nonce, PublicKey, PrivateKey, Deoxysii, encrypt, decrypt, } from '@oasislabs/confidential';
import { Web3Gateway } from '@oasislabs/web3';
import workspace from './workspace';
import Wallet from './wallet';
export default {
    Service: Service,
    Address: Address,
    Balance: Balance,
    deploy: deploy,
    Wallet: Wallet,
    setGateway: setGateway,
    defaultOasisGateway: defaultOasisGateway,
    gateways: {
        Gateway: Gateway,
        Web3Gateway: Web3Gateway,
    },
    utils: {
        DeployHeader: DeployHeader,
        Deoxysii: Deoxysii,
        OasisCoder: OasisCoder,
        Nonce: Nonce,
        PrivateKey: PrivateKey,
        PublicKey: PublicKey,
        bytes: bytes,
        cbor: cbor,
        decrypt: decrypt,
        encrypt: encrypt,
        keccak256: keccak256,
        idl: {
            fromWasm: fromWasm,
            fromWasmSync: fromWasmSync,
        },
    },
    workspace: workspace,
    disconnect: function () {
        try {
            defaultOasisGateway().disconnect();
        }
        catch (_e) {
            // `defaultOasisGateway` will throw if there's no default gateway.
            // Disconnecting from a nonexistent gateway is a no-op.
        }
    },
};
